backends/qualcomm/passes/convert_addmm_back_to_linear.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) Qualcomm Innovation Center, Inc.
```

Comment:
Please see my comments on quantization_tag. Youl should not need this hopefully.

Reply:
We still need this pass because it produces some view_copy ops after to_edge. For these ops, they seem not to have quantization_tag. And QNN HTP does not support keepdim, the view_copy(reshape) should exist for now.

Reply:
Can you elaborate a bit on keepdim? Which op keepdim is on?

> For these ops, they seem not to have quantization_tag

Can you give a concrete example, like with some code I can use to reproduce? Basically if you are tagging "linear" node, e.g. aten.linear, functional.linear, or torch._C.linear, you should be able to preserve the tag.

I think I know the issue in transformer cases, but just wanted to confirm. Is this for quantized linear or fp32 linear.

Reply:
You could use this test and change the dimension of input into three dimensions to reproduce.
``` 
def test_qnn_backend_ptq_linear(self):
        class Linear(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.linear = torch.nn.Linear(4, 5).eval()

            def forward(self, x):
                return self.linear(x)

        example_inputs = (torch.ones([2, 3, 4]),)
        instance = Linear()
        quant_instance = get_qdq_module(instance, example_inputs)
        buffer = self.lower_module_and_test_output(quant_instance, example_inputs)
        model_name = "ptq_qnn_linear_model"
        save_model_and_expected_output(instance, buffer, example_inputs, model_name)
```

https://github.com/pytorch/executorch/blob/533b1c871bee3d098340287e6eafb8b1cbbe1c40/backends/qualcomm/tests/test_qnn_delegate.py#L855

You could see two view copy in the graph after to_edge, so we need to annotate the output of the first view copy and linear op.
![image](https://github.com/pytorch/executorch/assets/138087975/20179872-fad7-46c8-836d-eb3d384c394a)

May I know do you have any other suggestion?

Reply:
Yes, you are right. The above figure is captured after convert addmm to linear.
https://github.com/pytorch/executorch/blob/533b1c871bee3d098340287e6eafb8b1cbbe1c40/backends/qualcomm/utils/utils.py#L68

I also think that they should share the same quantization_tag. But we annotate linear op with the following.
https://github.com/pytorch/executorch/blob/533b1c871bee3d098340287e6eafb8b1cbbe1c40/backends/qualcomm/qnn_quantizer.py#L542

When I run the linear example above, I get a key error when printing "node.meta['quant_attrs']" before the code below.
So, they seem not to have the same quantization_tag.
https://github.com/pytorch/executorch/blob/533b1c871bee3d098340287e6eafb8b1cbbe1c40/backends/qualcomm/passes/convert_addmm_back_to_linear.py#L64
![image](https://github.com/pytorch/executorch/assets/138087975/8de331f5-fac7-44d9-91e8-458ab3686529)

Reply:
I forget but are quant_attrs added to node meta before lowering to edge via `to_edge` which applies decomposition? If so I can understand why it is not preserved. But in that case I would re-evaluate when you should convert `dq -> linear -> q` to simply `linear` and add quant_attr meta on linear node. If you are doing this after quantization but before to_edge then yes, you will lose this information. 

I think we need to solve this better. I will stamp the diff for now, but lets file a task.

Reply:
Thanks for the clarification. 
In my view, quant_attrs are added after to_edge.
https://github.com/pytorch/executorch/blob/533b1c871bee3d098340287e6eafb8b1cbbe1c40/backends/qualcomm/utils/utils.py#L75
The AnnotateQuantAttrs pass will add quant_attrs into the source node of quantize node. So, it will rely on quantize node.

With respect to linear node, there is not quantize node after addmm and the first view copy. Therefore, we need this pass, convert_addmm_back_to_linear, to annotate quant_attr.
May I know is it expected that new node does not have Q & DQ nodes after quantization and decomposition, such as view_copy in this case?



Quantization Stage:
input -----> Q -----> DQ ¯¯\\
weight ---> Q -----> DQ  -> Linear -----> Q ------>DQ
bias ------> Q -----> DQ   __/

After to_edge:
input -----> Q -----> DQ  -----> permute_copy  ¯¯¯¯\\
weight ---> Q -----> DQ  -----> **view_copy**  ----> addmm -------> **view_copy** -----> Q ------>DQ
bias ------> Q -----> DQ   ___________________________/

After convert addmm to linear:
input -----> Q -----> DQ ¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯\\
weight ---> Q -----> DQ  -----> **view_copy**  ----> **Linear**-------> view_copy -----> Q ------>DQ
bias ------> Q -----> DQ   ______________________________/

After AnnotateQuantAttrs  and Fold_QDQ (Bold words indicate that quant_attr has been annotated)
**_input_**  ¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯\\
**_weight_**  -----> view_copy  ----> Linear-------> **_view_copy_** 
**_bias_**   ______________________________/

Reply:
> May I know is it expected that new node does not have Q & DQ nodes after quantization and decomposition, such as view_copy in this case?

That is correct. In fact it will be exactly how you have depicted, very nicely, in your diagram.

However, I have some concerns with the way you are identifying quantized linear and corresponding quantization attributes.
- With new update, `weight -> Q ->DQ -> linear` will be const propagated such that the model will not have fp32 weights and will only have quantized weights. Thus after `convert` step of quant workflow you will have `quantized_weight -> DQ -> linear`. So your transformations wont be valid.
- Also for correctness of lowering, I suggest keeping `DQ -> OP -> Q` pattern and match against that instead of doing "Fold Q/DQ". May I know why did you chose to do "fold q/dq".

Reply:
Thanks for your information.
We will discuss about it more.